### PR TITLE
Add `init fill-config-transformer` CLI command

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

This command reads the Hugging Face model name and revision from the `initialize.components.transformer.encoder_loader` config section, fetches its config from the HF Model Hub and fills in the entry point parameters for the same.

### Types of change

New feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: .github/workflows/test.yml
Comment (adrianeboyd) [inline .github/workflows/test.yml:79]: I'd recommend using `transformers[sentencepiece]` instead.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (danieldk) [inline spacy_curated_transformers/cli/fill_config_transformer.py:56]: Why not a regular enum?

Also, the last time I tried, `IntEnum` also works without these tupled-type members? The `IntEnum` documentation also uses regular integers?

Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Changed to `Enum`.

**Review by danieldk — APPROVED** [review]
LGTM, feel free to merge if @adrianeboyd is also ok with it.

File: .github/workflows/test.yml
Comment (adrianeboyd) [inline .github/workflows/test.yml:6]:
```suggestion
  hf-transformers-pip: transformers[sentencepiece]>=3.4.0,<4.32.0
```

File: spacy_curated_transformers/tests/test_cli_app.py
Comment (adrianeboyd) [inline spacy_curated_transformers/tests/test_cli_app.py:123]: Can we add checks for some of the expected values for the models above?

(Is the result `model_max_length = 9999` right for `explosion-testing/roberta-test`?)

**Comment by adrianeboyd** [comment]
I could run the CLI command for the prefilled config snippets, but there's still no way for users to get to an initial config:

```
nlp.add_pipe("curated_transformer")
```

or

```
spacy init config -p curated_transformer
```

both fail with:

```
Config validation error
curated_transformer.model -> vocab_size	field required
{'@architectures': 'spacy-curated-transformers.XlmrTransformer.v1', 'piece_encoder': {'@architectures': 'spacy-curated-transformers.XlmrSentencepieceEncoder.v1'}, 'with_spans': {'@architectures': 'spacy-curated-transformers.WithStridedSpans.v1'}}
```

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:251]: This seems potentially too low? Maybe int32.max so it's a more obvious max?

**Comment by shadeMe** [comment]
> I could run the CLI command for the prefilled config snippets, but there's still no way for users to get to an initial config:

Fixed in #20.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Done.

File: spacy_curated_transformers/tests/test_cli_app.py
Comment (shadeMe) [reply spacy_curated_transformers/tests/test_cli_app.py:307]: Added checks for the expected values.

**Comment by adrianeboyd** [comment]
I think that this command should include an option to fill in both the model loader and the piecer.

Right now (with the `vocab_size`) I get this far:

```
$ spacy init config -p curated_transformer config.cfg
$ spacy init fill-config-transformer config.cfg
ℹ Updating config for Curated Transformer pipe

'curated_transformer'

✘ Pipeline config does not have a valid model loader configuration for
the 'curated_transformer' Curated Transformer component
```

I feel like I should be able provide something like `xlm-roberta-base` as a command-line option to get the full `[initialize]` block? I don't think it needs to go beyond supporting HF models (similiar to the rest), but some way to get to a full config that I can then use with `spacy train`?

**Comment by shadeMe** [comment]
> I feel like I should be able provide something like xlm-roberta-base as a command-line option to get the full [initialize] block? I don't think it needs to go beyond supporting HF models (similiar to the rest), but some way to get to a full config that I can then use with spacy train?

Added.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:29]: I understand the motivation, but I think `"main"` would be a more helpful default. A warning about using `"main"` instead from our side would be fine?

Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:29]:
```suggestion
    model_revision: Optional[str] = Opt(None, "--model-revision", "-r", help="Revision of the Hugging Face model. Must be provided with the model name."),
```

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:188]: Probably to be modified, but:
```suggestion
            "Either both model name and model revision are provided as command-line arguments or neither of them are",
```

File: spacy_curated_transformers/tests/test_cli_app.py
Comment (adrianeboyd) [inline spacy_curated_transformers/tests/test_cli_app.py:338]: This is a bad sign?

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:248]: Tell people about the CLI model options in this error message.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [inline spacy_curated_transformers/cli/fill_config_transformer.py:131]: Give more pointers about what to do here, too, at the very least point to the explicit section of the config they need to look at. Probably better something like a link to the docs, since it may be too much for this message otherwise?

**Review by adrianeboyd — COMMENTED** [review]
I can get to full valid config now! I think the UX could still be improved a bit because users unfamiliar with the library probably won't have a good idea of where to start with all the config settings.

**Comment by adrianeboyd** [comment]
And I think `fill-config-transformer` is going to be a confusing name and make it not obvious it's for `curated_transformer` only. Not that there are a lot of good choices...

Let's see, to get started:

* `fill-curated-config`
* `fill-curated-transformer`

I don't like the way that `fill` doesn't mean quite the same thing as in `fill-config`, but the alternatives are probably worse.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (adrianeboyd) [reply spacy_curated_transformers/cli/fill_config_transformer.py:424]: I think `int32.max` would be okay?

Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Changed to `int32.max`.

File: spacy_curated_transformers/tests/test_cli_app.py
Comment (shadeMe) [reply spacy_curated_transformers/tests/test_cli_app.py:338]: Removed.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: I've expanded the error a bit to mention which config value they should look at. However, I didn't add a link as the docs PR is still in the works (hyperlinks/ids will most likely change before it's merged).

Comment (adrianeboyd) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Is there a way to identify the possible list of architectures from the registry and provide those?

Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: It might be possible, but I'm not sure if that alone would be informative enough to the user?

**Comment by adrianeboyd** [comment]
Can you tell the user based on the downloaded model which kind of architecture they probably need? (Like, I was even confused by the naming of something that looks like bert but wasn't.)

**Comment by shadeMe** [comment]
I've expanded the error message to display either the expected HF model type for the given curated transformers architecture (in case the type of the HF model is not supported) or the correct curated transformers architecture (if the HF model type is supported).

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:38]: Is this PR already created?

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:36]: Note that this description is used when you run `spacy init`:

> Commands:
>   config                    Generate a starter config file for training.
>   fill-config               Fill partial config file with default values.
>   fill-curated-transformer  Fetches the hyperparameters of a Curated...
>   labels                    Generate JSON files for the labels in the data.
>   vectors                   Convert word vectors for use with spaCy.

which might get truncated on a console, so if we can shorten this that'd be great.

**Review by svlandeg — COMMENTED** [review]
(still reviewing - first few comments)

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py:38]: This should part of the spaCy curated transformers docs PR - https://github.com/explosion/spaCy/pull/12677

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: I've shorted it, but it might still get truncated. Anything shorter will be incomplete, I think?

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:430]: Try running this with something like `--model-name foobar` and you'd get a long and ugly error stack trace. Can we catch that, provide a commandline-friendly message and explain to the user what this value should look like?

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Unless we'd want to abbreviate to "CT" just for this doc string...

Otherwsise: "Fill Curated Transformer config parameters" or "Fill Curated Transformer parameters" ?

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:292]:
```suggestion
            f"Hugging Face model of type '{incoming_hf_model_type}' cannot be loaded into "
            f"Curated Transformer pipe '{transformer_name}' of type '{expected_model_type}' - "
```

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:307]: I would structure this `if-else` and the error msg compilation slightly differently.
First, immediately after L288, check whether `incoming_hf_model_type` is supported, and if not, focus the error msg on that.

Otherwise, if it's valid/supported but the type is not the same as the `expected_model_type`, explain the incompatibility and recommend to either change the model name (according to the expected type) or change the architecture in the config.

Comment (svlandeg) [reply spacy_curated_transformers/cli/fill_config_transformer.py:307]: (this just feels easier to maintain to me, also avoiding the partly compiled error msg that is reused which makes it slightly more difficult to follow what exactly is being printed and when)

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:317]: I assume we don't really want to provide a default model name for each of the types in `CURATED_TRANSFORMER_TO_HF_MODEL_TYPE`, right? If not, could we print at least an example to showcase to the user what is needed? e.g. store a dict that maps `XlmrTransformer` to `xlm-roberta-base` as possible model name and suggest that?

**Comment by svlandeg** [comment]
> I don't like the way that fill doesn't mean quite the same thing as in fill-config, but the alternatives are probably worse.

I was actually thinking the same earlier. What about `update` instead of `fill`? 🤔

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:80]: `model_max_length` is keyed here twice.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:342]:
```suggestion
        assert value
        filled_params[param_name] = value
```

Comment (svlandeg) [inline spacy_curated_transformers/cli/fill_config_transformer.py:342]:
```suggestion
        assert value
        filled_params[param_name] = value
```

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py]: Shortened it to `Fill Curated Transformer parameters from HF Hub`.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py:80]: Removed.

**Comment by shadeMe** [comment]
> I was actually thinking the same earlier. What about update instead of fill? 🤔

I'd still keep `fill` as the grand majority of the fields being added are actually missing in the config, i.e., there are no keys for them (as opposed to updating the values of existing keys).

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py:430]: Added a `try...except` block to handle errors. The error message has to be somewhat under-specified as a number of things could potentially fail during this process (on HF Hub's end) is large.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py:306]: Done.

File: spacy_curated_transformers/cli/fill_config_transformer.py
Comment (shadeMe) [reply spacy_curated_transformers/cli/fill_config_transformer.py:316]: Hmm, I'm a fan of neither as we don't want to hardcode anything that's out of our control (and model/repo names are more likely to change than model types). I've added example model names to the error message, all the same.

**Review by svlandeg — APPROVED** [review]
Looks great!

**Pull request merged**
